hypothesis-python/src/hypothesis/internal/conjecture/providers.py

```diff
@@ -436,39 +491,55 @@ def draw_float(
         max_value: float = math.inf,
         allow_nan: bool = True,
         smallest_nonzero_magnitude: float,
-        # TODO: consider supporting these float widths at the IR level in the
-        # future.
-        # width: Literal[16, 32, 64] = 64,
-        # exclude_min and exclude_max handled higher up,
     ) -> float:
-        (
-            sampler,
-            clamper,
-            nasty_floats,
-        ) = self._draw_float_init_logic(
-            min_value=min_value,
-            max_value=max_value,
-            allow_nan=allow_nan,
+        assert self._random is not None
+
+        constraints = {
+            "min_value": min_value,
+            "max_value": max_value,
+            "allow_nan": allow_nan,
+            "smallest_nonzero_magnitude": smallest_nonzero_magnitude,
+        }
+        if (constant := self._maybe_draw_constant("float", constraints)) is not None:
+            return constant
+
+        # on top of the probability to draw a constant float, we independently
+        # upweight math.inf, -math.inf, math.nan, and boundary values.
+        weird_floats = [
+            f
+            for f in [
+                math.inf,
+                -math.inf,
+                math.nan,
+                min_value,
+                next_up(min_value),
+                min_value + 1,
+                max_value - 1,
+                next_down(max_value),
+                max_value,
+            ]
+            if choice_permitted(f, constraints)
+        ]
+
+        if weird_floats and self._random.random() < 0.03:
+            return self._random.choice(weird_floats)
+
+        clamper = make_float_clamper(
+            min_value,
+            max_value,
             smallest_nonzero_magnitude=smallest_nonzero_magnitude,
+            allow_nan=allow_nan,
         )
 
-        assert self._cd is not None
+        sign = 1 if self._random.getrandbits(1) else -1
+        result = sign * lex_to_float(self._random.getrandbits(64))
 
-        while True:
-            i = sampler.sample(self._cd) if sampler else 0
-            if i == 0:
-                result = self._draw_float()
-                if allow_nan and math.isnan(result):
-                    clamped = result  # pragma: no cover
-                else:
-                    clamped = clamper(result)
-                if float_to_int(clamped) != float_to_int(result) and not (
-                    math.isnan(result) and allow_nan
-                ):
-                    result = clamped
-            else:
-                result = nasty_floats[i - 1]
-            return result
+        clamped = result if allow_nan and math.isnan(result) else clamper(result)
+        if float_to_int(clamped) != float_to_int(result) and not (
+            math.isnan(result) and allow_nan
+        ):
+            result = clamped
+        return result
```

Comment:
This diff is inlining `_draw_float` and removing the unnecessary `while True`

Reply:
This doubly-stacked change was too confusing to follow, I'll do this refactoring some other time.